src/java/org/apache/cassandra/db/compaction/CompactionTask.java

```diff
@@ -188,26 +187,65 @@ protected void runMayThrow() throws Exception
         // it is not empty, it may compact down to nothing if all rows are deleted.
         assert transaction != null;
 
-        if (transaction.originals().isEmpty())
+        if (inputSSTables().isEmpty())
             return;
 
         if (DatabaseDescriptor.isSnapshotBeforeCompaction())
             realm.snapshotWithoutMemtable(System.currentTimeMillis() + "-compact-" + realm.getTableName());
 
         // The set of sstables given here may be later modified by buildCompactionCandidatesForAvailableDiskSpace() and
         // the compaction iterators in CompactionController and OverlapTracker will reflect the updated set of sstables.
-        try (CompactionController controller = getCompactionController(transaction.originals());
+        try (CompactionController controller = getCompactionController(inputSSTables());
              CompactionOperation operation = createCompactionOperation(controller, strategy))
         {
             operation.execute();
         }
     }
 
+    /**
+     * @return The token range that the operation should compact. This is usually null, but if we have a parallelizable
+     * multi-task operation (see {@link UnifiedCompactionStrategy#createAndAddTasks}), it will specify a subrange.
+     */
+    protected Range<Token> tokenRange()
+    {
+        return null;
+    }
+
+    /**
+     * If this is a partial compaction, its progress reports are shared between tasks. This method returns the shared
+     * progress object.
+     */
+    protected SharedCompactionProgress sharedProgress()
+    {
+        return null;
+    }
+
+    /**
+     * @return The set of input sstables for this compaction. This must be a subset of the transaction originals and
+     * must reflect any removal of sstables from the originals set for correct overlap tracking.
+     * See {@link UnifiedCompactionTask} for an example.
+     */
+    protected Set<SSTableReader> inputSSTables()
+    {
+        return transaction.originals();
+    }
```

Comment:
I see that not all `transaction.originals()` references in this class were replaced by `inputSSTables()`. Is that intentional? It seems like it might only be in methods that are not overridden in subclasses, but I worry that this might be fragile, assuming I understand it correctly.

Reply:
I updated them, changing everything that should use `inputSSTables`. Can you take another look?